### DISCUSSION TITLE
iOS finishTransaction

### DISCUSSION THREAD

**Discussion by sammiepls** [hyochan/react-native-iap#1674] [Q&A]
Hi there, I've been trying to implement this package into my app but have been facing issues  with iOS. If I open the subscription and then close the modal, I keep getting this error:

```
{"code": "E_USER_CANCELLED", 
"debugMessage": "The operation couldn’t be completed. (SKErrorDomain error 2.)", "message": "The operation couldn’t be completed. (SKErrorDomain error 2.)", 
"responseCode": "2"}
```
I am following the medium article about hooks and adding an extra `finally` to try and make sure that the transaction is finished but I'm still getting the error. Does anyone know what is the correct way to implement it? :

```
 const receipt = purchase.transactionReceipt;
        if (receipt) {
          try {
            const ackResult = await finishTransaction(purchase);
            console.log("ackResult", ackResult);
          } catch (ackErr) {
            console.warn("ackErr", ackErr);
          } finally {
            await clearTransactionIOS();
          }
 ```

**Comment by rohitkataria** [comment]
you can use this function inside purchaseErrorListener for ios case:
  purchaseErrorListner = purchaseErrorListener((error: PurchaseError) => {
        showError(error);
      });

function showError(error:any) {
  let errorMessage = '';
  if (typeof error === 'string') {
      errorMessage = error;
  } else if (error instanceof Error) {
      // If it's an error object, use its message property
      errorMessage = error.message;
  } else if (typeof error === 'object' && error.message) {
      // If it's an object and has a message property, use it
      errorMessage = error.message;
  } else {
      // Default message if none of the above
      errorMessage = 'An unexpected error occurred';
  }
  // Alert.alert('Error', errorMessage);
}
